### PR TITLE
actor critic algorithm

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
PR for issue: https://github.com/HorizonRobotics/alf/issues/342

### CODE REVIEW

File: alf/algorithms/actor_critic_algorithm.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 import gin.tf
 import tensorflow as tf
 
-from tf_agents.networks.network import Network, DistributionNetwork
+from alf.utils.networks import Network, DistributionNetwork
```
Comment (hnyu) [inline alf/algorithms/actor_critic_algorithm.py]: Now from alf.networks import ...

File: alf/utils/common.py [hunk 2]
```diff
@@ -28,6 +28,9 @@
 import time
 import random
 import numpy as np
+import torch
+from alf.utils.nest_utils import map_structure
```
Comment (hnyu) [inline alf/utils/common.py]: Now from alf.nest import ...

File: alf/utils/common.py [hunk 3]
```diff
@@ -583,50 +586,44 @@ def explained_variance(ypred, y):
     return 1 - tf.nn.moments(y - ypred, axes=(0, ))[1] / (vary + 1e-30)
 
 
-def sample_action_distribution(distributions, seed=None):
+def sample_action_distribution(nested_distributions):
```
Comment (emailweixu) [inline alf/utils/common.py]: let's move this to dist_utils

File: alf/utils/common.py [hunk 4]
```diff
@@ -583,50 +586,44 @@ def explained_variance(ypred, y):
...
     """
-
-    seed_stream = tfp.util.SeedStream(seed=seed, salt='sample')
-    return tf.nest.map_structure(lambda d: d.sample(seed=seed_stream()),
-                                 distributions)
+    return map_structure(lambda d: d.sample(), nested_distributions)
 
 
-def epsilon_greedy_sample(nested_dist, eps=0.1):
+def epsilon_greedy_sample(nested_distributions, eps=0.1):
```
Comment (emailweixu) [inline alf/utils/common.py]: move this too

File: alf/utils/common.py [hunk 5]
```diff
@@ -583,50 +586,44 @@ def explained_variance(ypred, y):
...
+    def greedy_fn(dist):
+        # pytorch distribution has no 'mode' operation
+        sample_action = dist.sample()
+        greedy_mask = torch.rand(sample_action.shape[0]) > eps
+        if isinstance(dist, torch.distributions.categorical.Categorical):
+            greedy_action = torch.argmax(dist.logits, -1)
+        elif isinstance(dist, torch.distributions.normal.Normal):
+            greedy_action = dist.mean
+        else:
+            greedy_action = dist.sample()
```
Comment (emailweixu) [inline alf/utils/common.py]: Let's raise NotImplementedError here

Comment (Haichao-Zhang) [reply alf/utils/common.py]: Changed to NotImplementedError

Comment (Haichao-Zhang) [reply alf/utils/common.py]: Changed to NotImplementedError

File: alf/utils/common.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/utils/common.py]: moved

File: alf/utils/common.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/utils/common.py]: moved

File: alf/algorithms/actor_critic_algorithm.py [hunk 6]
```diff
@@ -13,15 +13,12 @@
 # limitations under the License.
 """Actor critic algorithm."""
 
-import gin.tf
-import tensorflow as tf
-
-from tf_agents.networks.network import Network, DistributionNetwork
-
+import gin
+from alf.networks import ActorNetwork, ValueNetwork
```
Comment (hnyu) [inline alf/algorithms/actor_critic_algorithm.py]: ActorNetwork should be ActorDistributionNetwork

File: alf/algorithms/actor_critic_algorithm.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_algorithm.py]: Changed

File: alf/utils/common.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/utils/common.py]: Changed

File: alf/algorithms/actor_critic_algorithm.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_algorithm.py]: Changed

**Comment by Haichao-Zhang** [comment]
Now added unit test ```actor_critic_algorithm_test.py```, which can run successfully when including the dependencies from PR Trainer #380.

**Comment by Haichao-Zhang** [comment]
Updated some function interface tested based on #380.

File: alf/algorithms/actor_critic_algorithm.py [hunk 7]
```diff
@@ -37,8 +35,10 @@ class ActorCriticAlgorithm(OnPolicyAlgorithm):
     def __init__(self,
                  observation_spec,
                  action_spec,
-                 actor_network: DistributionNetwork,
-                 value_network: Network,
+                 env,
+                 config,
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_algorithm.py]: Since these two are optional. Do you think it's better to add `set_env()` and `set_config()` to RLAlgorithm?

File: alf/algorithms/actor_critic_algorithm_test.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+import alf
+from alf.utils import common, dist_utils, tensor_utils
+from alf.data_structures import StepType, TimeStep
+from alf.networks import ActorDistributionNetwork, ValueNetwork
+from alf.algorithms.rl_algorithm import RLAlgorithm
+from alf.algorithms.actor_critic_algorithm import ActorCriticAlgorithm
+
+
+# adapted from alf.algorithms.rl_algorithm_test
+class MyEnv(object):
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_algorithm_test.py]: Let's use the one from rl_algorithm_test directly

File: alf/algorithms/actor_critic_algorithm_test.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        state = alg.get_initial_predict_state(env.batch_size)
+        policy_step = alg.rollout_step(time_step, state)
+        logits = policy_step.info.action_distribution.logits
+        print("logits: ", logits)
+        self.assertTrue(torch.all(logits[:, 1] > logits[:, 0]))
+        self.assertTrue(torch.all(logits[:, 1] > logits[:, 2]))
+
+
+if __name__ == '__main__':
+    ActorCriticAlgorithmTest().test_ac_algorithm()
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_algorithm_test.py]: alf.test.main()

File: alf/algorithms/actor_critic_algorithm_test.py (hunk 8 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_algorithm_test.py]: Changed

File: alf/algorithms/actor_critic_algorithm_test.py (hunk 9 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_algorithm_test.py]: Changed

File: alf/algorithms/actor_critic_algorithm.py (hunk 7 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_algorithm.py]: Changed ```env``` and ```config``` to optional to be consistent with RLAlgorithm currently. The current way looks good to me. Changing to ```set``` function does not provide too much advantage.

File: alf/algorithms/actor_critic_algorithm.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_algorithm.py]: @hnyu changes have done according to your comments. Do you have any more comments?

File: alf/algorithms/actor_critic_algorithm.py [hunk 10]
```diff
@@ -47,18 +48,21 @@ def __init__(self,
         """Create an ActorCriticAlgorithm.
 
         Args:
+            observation_spec (nested TensorSpec): representing the observations.
             action_spec (nested BoundedTensorSpec): representing the actions.
-            actor_network (DistributionNetwork): A network that returns nested
+            env ():
+            config ():
```
Comment (hnyu) [inline alf/algorithms/actor_critic_algorithm.py]: more specific comments for these two new args? If the base algorithm has the comments then probably just copy them here.

File: alf/utils/dist_utils.py [hunk 11]
```diff
@@ -251,6 +251,47 @@ def _compute_log_prob(single_distribution, single_action):
     return total_log_probs
 
 
+def sample_action_distribution(nested_distributions):
```
Comment (hnyu) [inline alf/utils/dist_utils.py]: IF this is a general function API, suggest adding a flag for `rsample` if reparameterized is needed

Comment (Haichao-Zhang) [reply alf/utils/dist_utils.py]: ```rsample``` option added

File: alf/algorithms/actor_critic_algorithm.py (hunk 10 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_algorithm.py]: comments added
